server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java

```diff
@@ -230,376 +234,427 @@ private void onlyCreateIndex(final CreateIndexClusterStateUpdateRequest request,
+            final DiscoveryNodes nodes = currentState.nodes();
+            final Version createdVersion = Version.min(Version.CURRENT, nodes.getSmallestNonClientNodeVersion());
+            indexSettingsBuilder.put(IndexMetaData.SETTING_INDEX_VERSION_CREATED.getKey(), createdVersion);
+        }
+        if (indexSettingsBuilder.get(SETTING_NUMBER_OF_SHARDS) == null) {
+            indexSettingsBuilder.put(SETTING_NUMBER_OF_SHARDS, settings.getAsInt(SETTING_NUMBER_OF_SHARDS, 1));
+        }
+        if (indexSettingsBuilder.get(SETTING_NUMBER_OF_REPLICAS) == null) {
+            indexSettingsBuilder.put(SETTING_NUMBER_OF_REPLICAS, settings.getAsInt(SETTING_NUMBER_OF_REPLICAS, 1));
+        }
+        if (settings.get(SETTING_AUTO_EXPAND_REPLICAS) != null && indexSettingsBuilder.get(SETTING_AUTO_EXPAND_REPLICAS) == null) {
+            indexSettingsBuilder.put(SETTING_AUTO_EXPAND_REPLICAS, settings.get(SETTING_AUTO_EXPAND_REPLICAS));
+        }
+
+        if (indexSettingsBuilder.get(SETTING_CREATION_DATE) == null) {
+            indexSettingsBuilder.put(SETTING_CREATION_DATE, Instant.now().toEpochMilli());
+        }
+        indexSettingsBuilder.put(IndexMetaData.SETTING_INDEX_PROVIDED_NAME, request.getProvidedName());
+        indexSettingsBuilder.put(SETTING_INDEX_UUID, UUIDs.randomBase64UUID());
+
+        if (sourceMetaData != null) {
+            assert request.resizeType() != null;
+            prepareResizeIndexSettings(
+                currentState,
+                mappings.keySet(),
+                indexSettingsBuilder,
+                request.recoverFrom(),
+                request.index(),
+                request.resizeType(),
+                request.copySettings(),
+                indexScopedSettings);
+        }
+
+        Settings indexSettings = indexSettingsBuilder.build();
+        /*
+         * We can not check the shard limit until we have applied templates, otherwise we do not know the actual number of shards
+         * that will be used to create this index.
+         */
+        MetaDataCreateIndexService.checkShardLimit(indexSettings, currentState);
+        return indexSettings;
+    }
 
-                for (int shardId = 0; shardId < tmpImd.getNumberOfShards(); shardId++) {
-                    indexMetaDataBuilder.primaryTerm(shardId, tmpImd.primaryTerm(shardId));
-                }
+    /**
+     * Calculates the number of routing shards based on the configured value in indexSettings or if recovering from another index
+     * it will return the value configured for that index.
+     */
+    static int getIndexNumberOfRoutingShards(Settings indexSettings, @Nullable IndexMetaData sourceMetaData) {
+        int numTargetShards = IndexMetaData.INDEX_NUMBER_OF_SHARDS_SETTING.get(indexSettings);
+        final Version indexVersionCreated = IndexMetaData.SETTING_INDEX_VERSION_CREATED.get(indexSettings);
+        int routingNumShards;
```

Comment:
Can you make this `final` so the compiler complains if it isn't set in a branch?